tests/data/test_performance.py

```diff
@@ -182,3 +186,81 @@ def test_no_output_returns_conf(self):
+            pytest.param(ValueError, "metric_names", [[], lambda x: x], id="No metrics"),
+            pytest.param(ValueError, "metric_fn", [["metric"], "not_callable"], id="Not Callable"),
+            pytest.param(ValueError, "Reserved metric", [["metric", "Count"], lambda x: x], id="Count is reserved"),
+        ],
+    )
+    def test_generate_metrics_init_fails(self, errorType, errorStr, args):
+        with pytest.raises(errorType) as error:
+            undertest.MetricGenerator(*args)
+        assert errorStr in str(error.value)
+
+    def test_generate_metrics_init_correctly(self):
+        metric = undertest.MetricGenerator(["test_metric"], lambda data, names: {"test_metric": 1})
+        assert metric.metric_names == ["test_metric"]
+        assert metric(pd.DataFrame()) == {"test_metric": 1}
+
+    def test_generate_named_metrics(self):
+        metric = undertest.MetricGenerator(["metric1", "metric2"], lambda data, names: {name: 1 for name in names})
+        assert metric(pd.DataFrame(), ["metric2"]) == {"metric2": 1}
+        with pytest.raises(ValueError) as error:
+            metric(pd.DataFrame(), ["metric3"])
+        assert "metric3" in str(error.value)
+
+    def test_generate_metrics_with_kwargs(self):
+        def metric_fn(data, metric_names: list[str], *, special: int = 2):
+            return {name: special for name in metric_names}
+
+        metric = undertest.MetricGenerator(["metric1", "metric2"], metric_fn=metric_fn)
+        assert metric(pd.DataFrame(), ["metric1"], special=3) == {"metric1": 3}
+
+
+class TestBinaryMetricGenerator:
+    def test_binary_metric_generator_init(self):
+        metrics = undertest.BinaryClassifierMetricGenerator()
+        assert metrics.rho == 1 / 3
+        assert metrics.metric_names == ALL_STATS[1:]
+
+    @mock.patch(
+        "seismometer.data.performance.calculate_binary_stats",
+        return_value={"Accuracy": 0.8, "Sensitivity": 0.7, "Specificity": 0.6, "PPV": 0.5, "NPV": 0.4},
+    )
+    def test_binary_metric_generation(self, mock_stats):
+        metrics = undertest.BinaryClassifierMetricGenerator(rho=0.01)
+        data = pd.DataFrame({"fake": [0, 1, 0, 1], "data": [0.4, 0.5, 0.6, 0.7]})
+        result = metrics(data, ["Accuracy", "PPV"], target_col="TARGET", score_col="SCORE", score_threshold=0.2)
+        assert result == {"Accuracy": 0.8, "PPV": 0.5}
+        mock_stats.assert_called_once_with(data, "TARGET", "SCORE", 0.2, rho=0.01)
+
+
+class TestBinaryStats:
+    def test_calc_ci_maps_roc_values(self):
+        _, truth, output, _, _ = ci_testcase0()
+        data = pd.DataFrame({"truth": truth, "output": output})
+        expected = {
+            "Threshold": 70.0,
+            "TP": 1.0,
+            "FP": 0.0,
+            "TN": 2.0,
+            "FN": 1.0,
+            "Accuracy": 0.75,
+            "Sensitivity": 0.5,
+            "Specificity": 1.0,
+            "PPV": 1.0,
+            "NPV": 0.6666666666666666,
```

Comment:
Is this hinting at an 'issue'?  Should we round/truncate somewhere?
Honestly, I'm a tiny bit surprised this test passes across all the environments

Reply:
```python
        # TN / TN + FN
        npv = np.divide(fps[-1] - fps, (fps[-1] - fps) + (tps[-1] - tps))
        npv[np.isnan(npv)] = 1
```
This is just the existing npv calculation not getting rounded. I can add a round to the dataframe to put everything at 4 decimals.